sisyphus/hash.py

```diff
@@ -69,13 +69,20 @@ def get_object_state(obj):
         return args, state
 
 
-def sis_hash_helper(obj):
+def sis_hash_helper(obj, _visited=None, _add_to_visited=True):
     """
     Takes most object and tries to convert the current state into bytes.
 
     :param object obj:
+    :param _visited: (internal use)
+    :param _add_to_visited: (internal use)
     :rtype: bytes
     """
+    if _visited is None:
+        # keep ref to obj alive to avoid having the same id for different objs
+        _visited = {}  # id -> (bytes, obj)
+    if id(obj) in _visited:
+        return _visited[id(obj)][0]
```

Comment:
Since we are already tracking the visited nodes we could also throw a meaningful error if a Node is visited again before the hash is computed. This means that there is a circular dependency, I think right now this would simply result in a "RecursionError: maximum recursion depth exceeded" exception.
`_visited[id(obj)] = None` could be used to mark nodes which are no finished yet.

Reply:
Yes but do we really get into such cases? So far any such case would have resulted in RecursionError, as you said, for this function and many other functions as well. And I think RecursionError is also fine here. I think we don't need to make the logic more complex (and also slightly slower because of the extra checks) just to provide a nicer error message.